xonsh/imphooks.py

```diff
@@ -0,0 +1,78 @@
+"""Import hooks for importing xonsh source files. This module registers
+the hooks it defines when it is imported.
+"""
+from __future__ import print_function, unicode_literals
+import os
+import sys
+import builtins
+from importlib.machinery import ModuleSpec
+from importlib.abc import MetaPathFinder, SourceLoader
+
+from xonsh.tools import string_types
+from xonsh.execer import Execer
+
+class XonshImportHook(MetaPathFinder, SourceLoader):
+    """Implements the import hook for xonsh source files."""
+
+    def __init__(self, *args, **kwargs):
+        super(XonshImportHook, self).__init__(*args, **kwargs)
+        self._filenames = {}
+        self._execer = None
+
+    @property
+    def execer(self):
+        if hasattr(builtins, '__xonsh_execer__'):
+            execer = builtins.__xonsh_execer__
+            if self._execer is not None:
+                self._execer = None
+        elif self._execer is None:
+            self._execer = execer = Execer(unload=False)
+        else:
+            execer = self._execer
+        return execer
+    #
+    # MetaPathFinder methods
+    #
+    def find_spec(self, fullname, path, target=None):
+        """Finds the spec for a xonsh module if it exists."""
+        spec = None
+        path = sys.path if path is None else path
+        name = fullname.rsplit('.', 1)[-1]
+        fname = name + '.xsh'
+        for p in path:
+            if not isinstance(p, string_types):
+                continue
+            if fname not in os.listdir(p):
```

Comment:
I ran into trouble right away with this trying to list a nonexistent location, and later trying to list a zip archive.  Adding a check before this line worked for me:
```
            if not os.path.isdir(p):
                continue
```